### PR TITLE
Allow concurrent outputs

### PR DESCRIPTION
**Pull request by mxsdev**
Solves #289 

Allows multiple parts of the code to concurrently manage a cell's outputs without having to worry about executions:

<p align="center">

</p>

To do this, I've unified output handling to a new file `cell.ts`, which acts as a single source of truth for all output handling.

Changes are synchronized using a new `Mutex` utility class, which essentially acts as a dynamic promise queue that ensures mutations to outputs are done in sequence. I am open to alternative patterns for accomplishing this.

### The API

The important parts are mostly in `cell.ts`, which has a `NotebookCellOutputManager` class. It is expected that one is created per each cell, where this would be the new central location for dealing with that cell's outputs. In particular, it has the responsibility of generating outputs for each output type, based on the metadata of that cell.

The expectation is that all matters of outputs are dealt with through this class. For example, to toggle the annotations output, you would do `outputs.toggleOutput(OutputType.annotations)`.

As more complex output behavior is needed, this class should provide the internal tools to make this happen by extending the API. All of the race conditions should be dealt with.

### Persisting Output Data

There is a broader move here from sending the data to an output through the json, to persisting it with the cell's metadata. This is necessary since cells will potentially be re-rendered as the user e.g. opens the annotation window, toggles outputs, etc.

The solution given is very far from excellent, and I am all ears to suggestions. But hopefully, it should work for now. 

There is a data duplication/caching issue (same data is sent over the messaging channel as is persisted through metadata cache), and it is perhaps confusing how you set the metadata and then have to manually call a refresh. 

In the future, we should introduce a broader abstraction to handle/unify state synchronization and persistence in outputs. Bonus points if we can make generating `NotebookCellOutput` objects part of this as well. Interesting if we can learn from marquee/tangle here, since it's a very similar problem.

## Known issues:

 - [x] A user manually removing outputs is not handled. Ideally we can listen to an event for this, if not we have to figure it out lazily.
 - [x] Non-gRPC execution code is untouched
 - [x] gRPC execution code needs to be unified for vercel & deno outputs
 - [x] Deno output should hide embedded terminal
 - [x] No unit tests (yet)
 
Other known issues:

 - Deno output won't persist promoted (if you toggle annotations, promotion status gets reset)

### CODE REVIEW

**Review by degrammer — APPROVED** [review]
This looks amazing, LGTM, tested, it feels super smooth

File: src/extension/cell.ts [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+} from 'vscode'
+
+import { OutputType } from '../constants'
+import { CellOutputPayload, Serializer } from '../types'
+import { Mutex } from '../utils/sync'
+
+import { getAnnotations, replaceOutput, validateAnnotations } from './utils'
+
+export class NotebookCellManager {
+  private data = new WeakMap<NotebookCell, NotebookCellOutputManager>()
```
Comment (christian-bromann) [inline src/extension/cell.ts:20]: I would advocate to use actual private fields, e.g.
```suggestion
  #data = new WeakMap<NotebookCell, NotebookCellOutputManager>()
```

unless you need the object for unit testing purposes.

File: src/extension/kernel.ts [hunk 2]
```diff
@@ -258,13 +265,19 @@ export class Kernel implements Disposable {
     })
   }
 
-  public async createCellExecution(cell: NotebookCell): Promise<NotebookCellExecution> {
-    return this.#controller.createNotebookCellExecution(cell)
+  public async createCellExecution(cell: NotebookCell): Promise<RunmeNotebookCellExecution> {
+    return await this.cellManager.createNotebookCellExecution(cell)
+    // return this.#controller.createNotebookCellExecution(cell)
```
Comment (christian-bromann) [inline src/extension/kernel.ts:270]: still needed?
```suggestion
```

File: src/extension/provider/annotations.ts [hunk 3]
```diff
@@ -49,41 +45,7 @@ export class AnnotationsProvider implements NotebookCellStatusBarItemProvider {
...
-      ])
-    } catch (e: any) {
-      if (e.message.toString().includes('controller is NOT associated')) {
-        return this.handleNotebookKernelSelection()
-      }
-      window.showErrorMessage(e.message)
-    } finally {
-      exec?.end(true)
-    }
+    await (await this.kernel.getCellOutputs(cell)).toggleOutput(OutputType.annotations)
```
Comment (christian-bromann) [inline src/extension/provider/annotations.ts:48]: maybe easier to read:
```suggestion
    const outputCell = await this.kernel.getCellOutputs(cell)
    await outputCell.toggleOutput(OutputType.annotations)
```

File: src/utils/sync.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+/* Mutex provides mutex feature by building a promise chain.
+  const m = new Mutex();
+  await m.withLock(async () => {
+    // critical section
+  })
+
+  https://github.com/golang/vscode-go/blob/fa820d45d7b9217e949c21ffd13268e20c7e0ee4/src/utils/mutex.ts#L8
+*/
+export class Mutex {
+	private mutex = Promise.resolve()
```
Comment (christian-bromann) [inline src/utils/sync.ts:10]: Can we format the file, seems like the indentation is sometimes 8 and sometimes 2

**Review by christian-bromann — COMMENTED** [review]
Some comments, but overall LGTM 👍

File: src/extension/kernel.ts (hunk 2 above)
Comment (mxsdev) [reply src/extension/kernel.ts:270]: Nope, thanks for catching!

**Comment by sourishkrout** [comment]
Something for next week - not super high priority short-term: Tried to review this branch but notebooks won't load for me for now. Assuming the final touches leading up to v1 led to inconsistencies and merge conflicts.

**Comment by mxsdev** [comment]
@sourishkrout 

I rebased against main, this should be working now. (Problem was that TLS wasn't implemented on this branch)

**Comment by mxsdev** [comment]
Now integrated with the notebook terminal 👍 

Also changed the "Open Terminal" button to just be "Terminal," and it now acts as a toggle across all output modes. I think this is better UX, let me know your thoughts!

**Comment by sourishkrout** [comment]
Let's perhaps get all the remaining Notebook Terminal stuff merged & shipped before releasing this one into `edge` for a break-in period. Makes sense, @mxsdev?

**Comment by mxsdev** [comment]
@sourishkrout 

Sounds good to me!

**Pull request merged**